File: 序号/2140. 解决智力问题.py
class Solution:
    def mostPoints(self, questions: List[List[int]]) -> int: 
        '''questions[i] [points,skips] 解决当前问题，获得分数point，但后面skips个问题都不能回答'''
        n = len(questions)
        dp = [0]*(n+1) # dp[i] [i,n-1]的最大值
        for i in range(n-1,-1,-1):
            j = questions[i][1]+1+i
            # i不拿，就跟i+1相同，i拿了，就是当前值加上j的值
            dp[i]=max(dp[i+1],questions[i][0]+(dp[j] if j<n else 0)) 
        return dp[0]

File: 序号/2140. 解决智力问题.kt
class Solution {
  fun mostPoints(questions: Array<IntArray>): Long {
      // 从左往右遍历，遍历到i的时候无法知道从前面的哪个跳过来
      // 从右往左遍历，i的最大值就是i+skip+1+cur
      val n = questions.size
      val dp = LongArray(n+1)
      for(i in n-1 downTo 0){
          val (cur,skip) = questions[i]
          dp[i]=max(dp[i+1],cur+if(i+1+skip>=n) 0 else dp[i+1+skip])
      }
      return dp[0]
  }
}https://static.eudic.net/web/dict/star-off.png